src/components/Experience.ts

```diff
@@ -29,11 +29,54 @@ export class Experience<TGameStartr extends FullScreenPokemon> extends GeneralCo
      *
      * @param pokemon   The pokemon to give experience to.
      * @param experience   The amount of experience to give.
+     * @returns True if the pokemon leveled up, false otherwise.
      */
-    public gainExperience(pokemon: IPokemon, experience: number): void {
+    public gainExperience(pokemon: IPokemon, experience: number): boolean {
         pokemon.experience += experience;
         if (pokemon.experience >= this.gameStarter.equations.experienceStarting(pokemon.title, pokemon.level + 1)) {
             this.levelup(pokemon);
+            return true;
         }
+        return false;
+    }
+
+     /**
+      * Processes experience gain for one or more pokemon
+      *
+      * @param battleInfo   Information about the current battle.
+      * @param partyIsWipedText   Text to be displayed when the opposing pokemon faints.
+      *
+      */
+    public processBattleExperience(battleInfo: IBattleInfo, onComplete: () => void): void {
+        const experienceToGain =  this.gameStarter.equations.experienceGained(
+            battleInfo.teams.player, battleInfo.teams.opponent);
+
+        // BattleInfo should keep track of all pokemon the player has sent out in the battle and process
+        // them one by one if they have not fainted to allow for shared exp. The same logic will apply
+        // for an exp share where the pokemon holding the exp share is always considered to have been "sent out"
+        // The code below will be applied to each sent out in the battle
+        const battleText: (string | string[])[][] = [[battleInfo.teams.player.selectedActor.nickname,
+                                                      " gained ", experienceToGain.toString(), " experience!"]];
+        this.gameStarter.menuGrapher.createMenu("GeneralText");
+        const levelUp = this.gainExperience(battleInfo.teams.player.selectedActor, experienceToGain);
+        if (levelUp) {
```

Comment:
Generally, if you're creating variables within an `if` statement, that's a sign it should be split into a different (perhaps `private`) function. 

It's generally best to have a lot of smaller functions instead of one big one, with specific, informative names all around. That makes the code self-documenting and much easier to read.

+1 to https://en.wikipedia.org/wiki/Single_responsibility_principle